### DISCUSSION TITLE
write_pajek includes no layer information

### DISCUSSION THREAD

**Discussion by bdklahn** [mapequation/infomap#283] [General]
I've created a map, and done a bunch of . . .
`add_multilayer_intra_link` calls.

But `write_pajek` includes no "*Intra" section (nor any Multilayer additional information, at all.)

If that exporter doesn't support that, can you point out which one does?
Maybe `write_json`?
I guess I can test and see . . .

Perhaps I am just misunderstanding the scope/capabilities of infomap.

Would it work if I just used `add_multilayer_link`, but with the same layer id?
I'm wondering if since "Multilayer" is not optional in the pajek spec, if the Infomap object might not even initialize that member/attribute if something with 'multilayer' in the function name is never called.

Basically I have some semantic information which can allow grouping Vertices into layers. I was hoping to use the Infomap functions to integrate this with some binary relationship information (e.g. which can be used to create edges).

**Comment by bdklahn** [comment]
Ok. I've confirmed that using `add_multilayer_link(source_multilayer_node, target_multilayer_node, weight=1.0)` doesn't seem to help in getting `write_pajek` to output any Mulitlayer information.

So I'm now seeing if `write_json` will contain that. But, so far, `write_json` has been running for like 10 minutes now.
And I'm fearful that the extra time is actually being spent on writing indentation spacing, and adding key names to every single node and layer id, every place they are used, while still not including any multilayer information.

Can someone suggest an output format which was implemented to include the layer information?

**Comment by bdklahn** [comment]
After maybe 15 minutes, `write_json` gave me this.
```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [22], in <module>
      1 module_layers_outpath_json = module_layers_outpath_pajek.with_suffix('.json')
----> 2 im.write_json(str(module_layers_outpath_json))

File /usr/local/lib64/python3.8/site-packages/infomap.py:6650, in Infomap.write_json(self, filename, states)
   6636 def write_json(self, filename, states=False):
   6637     """Write result to a JSON file.
   6638 
   6639     See Also
   (...)
   6648         If the state nodes should be included. Default ``False``.
   6649     """
-> 6650     return self.writeJsonTree(filename, states)

File /usr/local/lib64/python3.8/site-packages/infomap.py:3303, in InfomapBase.writeJsonTree(self, *args)
   3302 def writeJsonTree(self, *args):
-> 3303     return _infomap.InfomapBase_writeJsonTree(self, *args)

RuntimeError: std::bad_alloc

```
Is that function going to need more than the ~160G of memory (and 3.6P of disk space) I have free?

**Comment by bdklahn** [comment]
Looks like I've got . . .
```bash
> infomap --version
Infomap version 2.1.0 compiled with OpenMP
See www.mapequation.org for terms of use.
```
I guess I'll see if I can get Pypi v 2.2.0, and see if any updates fixes things for me.

**Comment by antoneri** [comment]
Hi and thanks for reaching out!

`write_pajek` writes the internal network representation that Infomap uses.

`write_json` writes the resulting clusters, mainly used in our web apps to simplify parsing.
You have discovered a bug where if you try to print the results before running the optimizer (`im.run()`), Infomap crashes.
(Also, since Infomap 2.0.2, the JSON output is minified without whitespace.)

Why do you want to print the pajek file? It's not needed to run Infomap if you use the Python module.
If you want a pajek file for use in another library, it's best that you generate it yourself. Infomap currently doesn't support generating its own input files (except its internal network representation).

I'll move this to a discussion post and create a separate issue for the write_json bug for now.

Please let us know if you have any questions.

**Reply by bdklahn** [reply]
Thanks for the response, @antoneri , and thanks for your work on this open software!

I was partly using the pajek output to confirm that the functions I was using to add layer information, were actually doing what I hoped/expected.
I found the functions on this page:
[https://mapequation.github.io/infomap/python/infomap.html](https://mapequation.github.io/infomap/python/infomap.html)
I was hoping something like . . .
`add_multilayer_intra_link(layer_id, source_node_id, target_node_id, weight=1.0)`
. . . would add to ' the internal network representation that Infomap uses', and thus be apparent in the pajek output.
That is, I was hoping then to see things like *Mulitlayer, in there. But I don't see that.
Maybe, for that, I also need to do `im.run()`, as well? I'm pretty sure did try that, at least once, before writing to pajek.

What we are trying to do . . .
We have a complex biologicl reaction to medical treatment. Many of our (potential) molecular "actors", are present in this large mapping model:
[https://acsn.curie.fr/ACSN2/downloads.html](https://acsn.curie.fr/ACSN2/downloads.html)
I was hoping to add the mutual membership in different 'maps' (and maybe 'modules') as InfoMap intra-layer relationships. We hoped to see some pattern arise where our 'actors' (maybe "highlighted" by up-weighting any relationships they are involved in, and/or otherwise labeling or filtering them) maybe tend to participate in interactions within or between certain layers. This is a bit of a discovery mission, but it seemed like Infomap might help shed some light.

If the layer information is not apparent in the pajek output, could you point me to the/a recommended way to see that?

**Reply by antoneri** [reply]
Alright, well almost! Infomap represents multilayer networks using "state nodes".
An explanation on this is available [here](https://www.mapequation.org/infomap/#InputMultilayer). There are examples of multilayer networks and if you continue reading until "States", the equivalent state network is shown.

You can inspect the (relaxed) state network using something like this:
```python
>>> from infomap import Infomap
>>> im = Infomap(silent=True, no_infomap=True)
>>> im.add_multilayer_intra_link(1, 1, 2)
>>> im.add_multilayer_intra_link(2, 1, 3)
>>> im.run()
>>> im.write_state_network("example.net")
>>> !cat example.net
# v2.2.0
# ./Infomap  --no-infomap --silent
*States
# stateId physicalId
0 1
1 2
2 3
3 1
*Links
0 1 0.925
0 2 0.075
1 0 1
2 3 1
3 1 0.075
3 2 0.925
```

But my suggestion is to not focus so much on the internal representation, but instead study the "[state tree](https://www.mapequation.org/infomap/#OutputTree)" output file which includes the layer id when running on multilayer networks.

Or, you can use the Python API:
```python
>>> from infomap import Infomap
>>> im = Infomap(silent=True)
>>> im.read_file("multilayer.net")
>>> im.run()
>>> for node in im.nodes:
...     print(node.node_id, node.module_id, node.layer_id)
...
1 1 2
2 1 2
3 1 2
1 2 1
4 2 1
5 2 1
``` 

We refer to clusters of nodes as "modules" (comes from the map equation using a modular description of a random walk on the network). Just so we avoid any confusion.

**Comment by bdklahn** [comment]
@antoneri , Thanks for taking the time to respond with all the useful information and examples!
I hope to continue with this tomorrow afternoon.

**Comment by bdklahn** [comment]
I don't know if this is the best place (topic, under which) to continue asking things.
And I had already read the previously cited page, which included the state tree information and other things, a few times, before I initially posted the Issue. But it wasn't clear to me, so that's why I just started experimenting with the Python package.
It seems I do now have ". . . multilayer_intra_link"s added into my map.
I've upweighted the links where our entities participate to 2, to attempt "highlight" them in the map.
Now I'm trying to figure out how to tune the system to get a tractable, yet interesting, result. We estimate a useful and interesting number of clusters might be something around 40. From reading . . . as I understand, that might have something to do with tuning the probabilities of "jumps/travels" between layers (inter). There are some seemingly relevant parameters I see in run function of the [API docs](https://mapequation.github.io/infomap/python/infomap.html). But It makes it hard to decide what to use and set, when the defaults have not been listed (presumably those could be auto-documented, if Python type hints were implemented, I think). I also see no information on "im??" introspection in Jupyter Lab (in fact it seems to cause the kernel to enter an infinite loop, or something). But maybe none of that "automatic" information is even possible if this is, like, SWIG, binding vs. pure Python.

Anyway, I was a little hopeful and surprised to see the 'preferred_number_of_modules' parameter, as a possible easy way to get to our target. But when I set that to 40, for example, I see things like . . .
'Per level number of modules:         [      19668,           0] (sum: 19668)'
. . . in the output. So is that not doing what I hoped it would do? Maybe this is not the way I'm supposed to go about this?

**Reply by antoneri** [reply]
Yes, this is a good place to ask such questions!

You are right that there are some drawbacks to using SWIG. We have a small Python wrapper around the swig exports to help out, but it's not perfect.

Don't worry about tuning the parameters until you get at least halfway "there". The modeling is the most important part. 
This includes `preferred_number_of_modules`.

From what I can tell, you get singleton modules: almost all nodes are alone in separate modules. Is this correct? This usually happens when you have a disconnected or extremely sparse network.

Usually, the best course of action is to draw a small schematic example of how your entities relate to each other and how they are stratified across different layers and assign meaningful link weights. You can even omit inter-layer links if you don't have a meaningful way to encode them. Infomap will then use a tunable relax rate (`multilayer_relax_rate = 0.15`) to jump between layers. You can run Infomap on your small example to verify that it works and you find the modules you expect.

The goal at this point is to get a modular description that's not just "put everything in one module" or "put all nodes in separate modules". Then you can tune your model to have layer-dependent relax rates using the inter-layer links.

Try it and let me know how it works!

**Reply by bdklahn** [reply]
Thanks so much for the response, @antoneri
It seems like  `multilayer_relax_rate` might be useful.
We were hoping Infomap could help inform what our model should be (narrowed down to). If we knew that, we might not even need Infomap. We could, say, go directly to creating and using our [COPASI](https://copasi.org) software, or something.
Every node I create in the model is connected to something, otherwise it wouldn't be in there. The map is literally generated from only connected things.
We have a list of 14091 binary relationships from a map created from cancer models, etc.
[https://acsn.curie.fr/navicell/maps/acsn2/master/index.html](https://acsn.curie.fr/navicell/maps/acsn2/master/index.html)
[https://acsn.curie.fr/ACSN2/downloads.html](https://acsn.curie.fr/ACSN2/downloads.html).

I create a link for every one. If both of any of those participate in any one of ~5 biological "maps", that link is an intra-layer link. I create no explicit inter-layer links. But I guess, implicitly, all other links will be inter-layer.
We have a set of 25 items (gene products) which are of interest to our specific study. So to "filter" the importance of those within this large connected map I just up-weight any of the links those are involved in.
So it seems like there is no "small schematic" we can draw which can be useful to the complexity we need to reduce.

I'm also wondering . . .
if I should be doing `include_self_links = False`
if there is a better way to "filter" only 25 nodes map out of this, besides using edge weights. I mean . . . I'm not sure how to keep in the Navicell nodes which have no direct connection to our 25 "actors of interest" in there.
Sorry . . . this stuff is a bit new to me. I did read the main Infomap paper, but . . .

**Reply by antoneri** [reply]
Just to clarify, by modeling I mean if you should represent your data as a regular "single layer" or multi-layer network, what the weights should be, what it means for a random walker to transition between layers, and how likely that should be. And so on... :)

As you might already know, Infomap finds flow modules that trap a random walk for a relatively long time. This is why network representation matters so much for community detection.

Since Infomap 2.0, we include self-links by default. You can omit them with `no_self_links`.

I'm not familiar with all your terms, so I can only help with what I think I recognize. I'm however of the opinion that it's best to get some results with the simplest possible model to verify that there's any meaningful community structure in your data.

**Reply by bdklahn** [reply]
Am I wrong in assuming I can induce communities, to test to see if I can get any modules (not a higher number than my total number of links), by increasing the weights on certain links?
```python
im.run(
    include_self_links=False,
    preferred_number_of_modules=40,
    multilayer_relax_rate = 0.15,
)
```
Here's my run . . .

```
===================================================
  Infomap ends at 2022-04-05 19:12:20
  (Elapsed time: 1.6s)
===================================================
=======================================================
  Infomap v2.2.0 starts at 2022-04-05 19:19:08
  -> Input network: 
  -> No file output!
  -> Configuration: preferred-number-of-modules = 40
=======================================================
  OpenMP 201511 detected with 40 threads...
  -> Warning: Higher order network specified but no higher order input found.
  -> Ordinary network input, using the Map Equation for first order network flows
Calculating global network flow using flow model 'undirected'... 
  -> Using undirected links.
  => Sum node flow: 1, sum link flow: 1
Build internal network with 28180 nodes and 8512 links...
Calculating one-level codelength... done!
 -> One-level codelength: 13.6768766
Calculating entropy rate... done!
  -> Entropy rate: 0

================================================
Trial 1/1 starting at 2022-04-05 19:19:08
================================================
Two-level compression: -1.4e+05% 0% 
Partitioned to codelength -2.34406985e-11 + 1 + 19628 = 19629 in 19668 (8512 non-trivial) modules.
Super-level compression: to codelength 19629 in 19668 top modules.

Recursive sub-structure compression: 0% . Found 2 levels with codelength 1

=> Trial 1/1 finished in 0.49s with codelength 1


================================================
Summary after 1 trial
================================================
Best end modular solution in 2 levels:
Per level number of modules:         [      19668,           0] (sum: 19668)
Per level number of leaf nodes:      [          0,       28180] (sum: 28180)
Per level average child degree:      [      19668,     1.43278] (average: 8085.41)
Per level codelength for modules:    [0.000000000, 0.000000000] (sum: 0.000000000)
Per level codelength for leaf nodes: [0.000000000, 1.000000000] (sum: 1.000000000)
Per level codelength total:          [0.000000000, 1.000000000] (sum: 1.000000000)
```




I only created 14091 links. In this case I made the link weights of interest 9.

**Reply by antoneri** [reply]
Now I think I know what's happening!

Are you using `im.add_node`? Try replacing it with `im.set_name` or removing the calls altogether.

I also see this line:
> Build internal network with 28180 nodes and 8512 links...

Something is wrong if you added 14000 links. Maybe the same cause as above.

How many nodes/layers are you adding?

**Reply by bdklahn** [reply]
I realize I actually _am_ still using `im.add_node`. I thought my code was currently only implicitly adding nodes by adding links. This is certainly a bug in my code, and will at least cause one major issue: I will be adding duplicate nodes where they participate in more than one link (which I am _depending_ on being true in _many_ cases).
Yes, it makes sense to either use the Infomap functionality to only create new nodes where they already exist, or create those name to id mappings myself, only when they don't already exist, using `dict().setdefault`.
I'll report back about this.

Thanks, @antoneri !

Also, while I'm here, I've been considering trying [NetworkX](https://networkx.org). It looks like that has some handy metadata, and adjacency, functionality, we might take advantage of. I also like that they can use strings (any hashable type) for node ids. Do you have any experience with that software. I'm not sure if, maybe, it is mostly pure Python, and might not be fast enough for large networks. I wonder about export/import to Infomap or better visualization tools.
Any thoughts or opinions?

Ah . . . I just noticed `add_networkx_graph(g, weight='weight')`. . .

**Comment by bdklahn** [comment]
This looks much better to me!

```
===================================================
  Infomap ends at 2022-04-06 14:51:14
  (Elapsed time: 1.824986s)
===================================================
=======================================================
  Infomap v2.2.0 starts at 2022-04-06 14:51:35
  -> Input network: 
  -> No file output!
  -> Configuration: flow-model = undirected
                    preferred-number-of-modules = 40
=======================================================
  OpenMP 201511 detected with 40 threads...
  -> Warning: Higher order network specified but no higher order input found.
  -> Ordinary network input, using the Map Equation for first order network flows
Calculating global network flow using flow model 'undirected'... 
  -> Using undirected links.
  => Sum node flow: 1, sum link flow: 1
Build internal network with 2774 nodes and 8512 links...
Calculating one-level codelength... done!
 -> One-level codelength: 10.3127556
Calculating entropy rate... done!
  -> Entropy rate: 2.562165122

================================================
Trial 1/1 starting at 2022-04-06 14:51:35
================================================
Two-level compression: -2.7e+03% 5.3e-13% 
Partitioned to codelength -6.61084823e-14 + 10.1874268 + 281 = 291.1874268 in 321 (30 non-trivial) modules.
Super-level compression: to codelength 291.1874268 in 321 top modules.

Recursive sub-structure compression: 26.8224918% 0.105587208% 0% . Found 4 levels with codelength 7.447033663

=> Trial 1/1 finished in 1.084034s with codelength 7.44703366


================================================
Summary after 1 trial
================================================
Best end modular solution in 4 levels:
Per level number of modules:         [        321,          40,          40,           0] (sum: 401)
Per level number of leaf nodes:      [          0,         405,        2250,         119] (sum: 2774)
Per level average child degree:      [        321,     1.38629,       57.25,       2.975] (average: 74.0518)
Per level codelength for modules:    [0.000000000, 1.271372682, 0.123354132, 0.000000000] (sum: 1.394726814)
Per level codelength for leaf nodes: [0.000000000, 0.020349875, 5.862130810, 0.169826164] (sum: 6.052306850)
Per level codelength total:          [0.000000000, 1.291722557, 5.985484942, 0.169826164] (sum: 7.447033663)
```

I actually kept `im.add_node`, but only created one if it wasn't already in my name to id dictionary.

I've been reading the docs, but I'm not completely clear on how to show those top "(30 non-trivial) modules".

Also, how might I filter for only the modules which include at least one of our nodes of interest?

Thanks so much for your time!

**Reply by antoneri** [reply]
It would be best if you did not use `im.add_node` in a multilayer context. If you need node names, use only `im.set_name` or skip it altogether!

This is why you get this warning:
> Warning: Higher order network specified but no higher order input found.

I'm telling you: don't do it. Your results are wrong if you do.

You can get the modules using `im.get_modules()`.

**Reply by bdklahn** [reply]
Ah . . . Gotcha. So there is more than one issue. Yeah . . . I was mainly using it, assuming the graph would be most authoritative in terms of what node ids were already in use. With previous ones added, I figured `im.num_nodes` would be a good way to get the next available node id I could use.
Ok. So I should be able to also pre-generate those from my external name to node id dictionary I create/use.
Thanks again.

**Reply by bdklahn** [reply]
I removed `im.add_node` and still get that warning.

```
=======================================================
  Infomap v2.2.0 starts at 2022-04-07 19:35:00
  -> Input network: 
  -> No file output!
  -> Configuration: flow-model = undirected
                    preferred-number-of-modules = 40
=======================================================
  OpenMP 201511 detected with 40 threads...
  -> Warning: Higher order network specified but no higher order input found.
  -> Ordinary network input, using the Map Equation for first order network flows
Calculating global network flow using flow model 'undirected'... 
  -> Using undirected links.
  => Sum node flow: 1, sum link flow: 1
Build internal network with 2483 nodes and 8512 links...
Calculating one-level codelength... done!
 -> One-level codelength: 10.3127556
Calculating entropy rate... done!
  -> Entropy rate: 2.562165122

================================================
Trial 1/1 starting at 2022-04-07 19:35:00
================================================
Two-level compression: 7.5% 11% 2.4672441% 0.0917344169% 0.0899256554% 0.0299058671% 0.32263725% 0.0392700562% 0.0126677936% 0.00166727441% 0.0477508972% 0.00418764572% 0.00178868898% 0.0157738289% 
Partitioned to codelength 0.617358273 + 7.59197093 + 0 = 8.209329201 in 40 modules.

Recursive sub-structure compression: 8.80478753% 0% . Found 3 levels with codelength 7.486515207

=> Trial 1/1 finished in 1.79508s with codelength 7.48651521


================================================
Summary after 1 trial
================================================
Best end modular solution in 3 levels:
Per level number of modules:         [         40,         360,           0] (sum: 400)
Per level number of leaf nodes:      [          0,         387,        2096] (sum: 2483)
Per level average child degree:      [         40,      18.675,     5.82222] (average: 9.62664)
Per level codelength for modules:    [0.617358273, 1.961671617, 0.000000000] (sum: 2.579029890)
Per level codelength for leaf nodes: [0.000000000, 1.548685951, 3.358799367] (sum: 4.907485317)
Per level codelength total:          [0.617358273, 3.510357568, 3.358799367] (sum: 7.486515207)
```

But it does look like I might be now getting the 40 modules I requested.

I only use . . .
`im.add_multilayer_intra_link`
and
`im.add_link`

Do you think I should replace `im.add_link` with `im.add_multilayer_inter_link` to get rid of that warning?
I guess that won't work because I'll have no idea what layers to use.
Maybe it doesn't really make sense to specify layers, if I can't provide any inter-layer links???
